subprojects/core/src/main/java/org/gradle/api/internal/FeaturePreviews.java

```diff
@@ -17,18 +17,39 @@
 
 import org.gradle.StartParameter;
 
+import java.util.EnumSet;
+
 public class FeaturePreviews {
-    private StartParameter startParameter;
+    private enum Feature {
+        ADVANCED_POM_SUPPORT,
+        GRADLE_METADATA
+    }
+
+    private EnumSet<Feature> enabledFeatures = EnumSet.noneOf(Feature.class);
 
     public FeaturePreviews(StartParameter startParameter) {
-        this.startParameter = startParameter;
+        if (startParameter.isAdvancedPomSupport()) {
+            enabledFeatures.add(Feature.ADVANCED_POM_SUPPORT);
+        }
+        if (startParameter.isGradleMetadata()) {
+            enabledFeatures.add(Feature.GRADLE_METADATA);
+        }
     }
 
-    public boolean isAdvancedPomSupportEnabled() {
-        return startParameter.isAdvancedPomSupport();
+    public void enableFeature(String name) {
+        try {
+            enabledFeatures.add(Feature.valueOf(name));
+        } catch (IllegalArgumentException e) {
+            // Ignore unknown features
```

Comment:
I think we should fail-fast when an attempt is made to enable an unknown feature.

Reply:
But that means removing a feature between versions would cause upgrades to fail the build. I thought the whole point here was to not make build changes mandatory.
However I think we should at least warn instead of plain ignore.

Reply:
Yeah good point. I guess we _could_ remember old features and ignore them when they become enabled by default. So enabling `GRADLE_METADATA` in 5.0 would be a no-op, since it is enabled by default.
But I don't feel strongly about it. The only reason _not_ to ignore would be to help users when they mis-spell.

Reply:
Had a chat with @melix and the failure is now immediate. `Feature` also have a boolean that indicates whether they are active or not, and enabling a feature that is not active is simply a no-op.